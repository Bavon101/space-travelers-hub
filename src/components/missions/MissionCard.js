import React from 'react';
import PropTypes from 'prop-types';

export default function MissionCard(props) {
  const { mission, id, description } = props;
  return (
    <tr>
      <td>
        {mission}
      </td>
      <td>
        {description}
      </td>
      <td>
        <div className="status-container">
          <h6>
            {id}
          </h6>
        </div>
      </td>
      <td>
        <button type="button" className="mission-btn">
          Join Mission
        </button>
      </td>
    </tr>
  );
}
MissionCard.propTypes = {
  mission: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
};
