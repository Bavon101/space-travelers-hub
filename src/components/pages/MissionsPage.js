import React from 'react';

export default function MissionsPage() {
  return (
    <div>
      <table>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th>action</th>
        </tr>
        <tr>
          <td>
            Thaicom
          </td>
          <td>
            The Missions section displays a list of current missions along with their brief
            description and participation status.
            There is also a button next to each mission that allows users to join the
            selected mission or leave the mission the user joined earlier.
          </td>
          <td>
            <div>
              <h6>
                not a member
              </h6>
            </div>
          </td>
          <td>
            <button type="button">
              Join Mission
            </button>
          </td>
        </tr>
      </table>
    </div>
  );
}
