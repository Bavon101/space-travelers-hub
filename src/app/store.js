import { combineReducers, createStore, applyMiddleware } from '@reduxjs/toolkit';
import thunk from 'redux-thunk';
import { missionsReducer } from '../components/missions/redux/missions';
import rocketReducer from '../components/rockets/redux/rockets';
import fetchRocketThunk from '../components/rockets/redux/rocketThunk';

const rootReducer = combineReducers({
  missionsReducer,
  rocketReducer,
});

const store = createStore(rootReducer, applyMiddleware(thunk));
store.dispatch(fetchRocketThunk());

export default store;
