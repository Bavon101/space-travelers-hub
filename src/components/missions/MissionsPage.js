import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import fetchMissions from '../../api/data_fetch';
import MissionCard from './MissionCard';
import './missions.css';

export default function MissionsPage() {
  const missions = useSelector((state) => state.missionsReducer);
  const dispatch = useDispatch();
  useEffect(() => {
    if (missions.length === 0) {
      dispatch(fetchMissions());
    }
  }, []);
  if (missions.length === 0) {
    return (<h3 className="no-missions-text">No missions</h3>);
  }
  return (
    <div>
      <table>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>

          </tr>
        </thead>
        <tbody>
          {
            missions.map((mission, index) => (
              <MissionCard
                key={mission.id}
                mission={mission.mission}
                id={mission.id}
                description={mission.description}
                index={index}
                reserved={mission.reserved}
              />
            ))
          }
        </tbody>
      </table>
    </div>
  );
}
