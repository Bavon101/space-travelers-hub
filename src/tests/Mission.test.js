import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../app/store';
import MissionCard from '../components/missions/MissionCard';
import MissionsPage from '../components/missions/MissionsPage';
import missioMockData from './mock_data/missions_mock_data';

beforeEach(() => {
  jest.spyOn(global, 'fetch').mockResolvedValue({
    json: jest.fn().mockResolvedValue(missioMockData),
  });
});

afterEach(() => {
  jest.restoreAllMocks();
});

it('Render Missions Page', () => {
  const missionsPage = renderer.create(
    <Provider store={store}>
      <MissionsPage />
    </Provider>,
  ).toJSON();
  expect(missionsPage).toMatchSnapshot();
});

it('Render Mission Card', () => {
  const missionCard = renderer.create(
    <Provider store={store}>
      <MissionCard
        mission={missioMockData[0].mission_name}
        id={missioMockData[0].mission_id}
        description={missioMockData[0].description}
        index={0}
        reserved={false}
      />
    </Provider>,
  ).toJSON();
  expect(missionCard).toMatchInlineSnapshot(`
<tr
  id="9D1B7E0"
  style={
    Object {
      "backgroundColor": "#DEE2E6",
    }
  }
>
  <td>
    <h3>
      Thaicom
    </h3>
  </td>
  <td>
    Thaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited, which is the company that owns and operates the Thaicom satellite fleet and other telecommunication businesses in Thailand and throughout the Asia-Pacific region. The satellite projects were named Thaicom by the King of Thailand, His Majesty the King Bhumibol Adulyadej, as a symbol of the linkage between Thailand and modern communications technology.
  </td>
  <td>
    <div
      className="status-container"
      style={
        Object {
          "backgroundColor": "gray",
        }
      }
    >
      <h6>
        NOT A MEMBER
      </h6>
    </div>
  </td>
  <td>
    <button
      className="mission-btn"
      onClick={[Function]}
      style={
        Object {
          "border": null,
          "color": null,
        }
      }
      type="submit"
    >
      Join Mission
    </button>
  </td>
</tr>
`);
});

test('Join a mission ', () => {
  const { container } = render(
    <Provider store={store}>
      <MissionCard
        mission={missioMockData[0].mission_name}
        id={missioMockData[0].mission_id}
        description={missioMockData[0].description}
        index={0}
        // render with reserved to false
        reserved={false}
      />
    </Provider>,
  );
  const reserveBtn = screen.getByText('Join Mission');
  act(() => {
    userEvent.click(reserveBtn);
  });
  // expect to be still Join Mission since the store MissionsReducer has no element hence no change
  expect(container.getElementsByClassName('mission-btn')[0].textContent).toBe('Join Mission');
});

test('Leave a mission ', () => {
  const { container } = render(
    <Provider store={store}>
      <MissionCard
        mission={missioMockData[0].mission_name}
        id={missioMockData[0].mission_id}
        description={missioMockData[0].description}
        index={0}
        // render with reserved to true
        reserved
      />
    </Provider>,
  );
  const reserveBtn = screen.getByText('Leave Mission');
  act(() => {
    userEvent.click(reserveBtn);
  });
  // expect to be still Leave Mission since the store MissionsReducer has no element hence no change
  expect(container.getElementsByClassName('mission-btn')[0].textContent).toBe('Leave Mission');
});
