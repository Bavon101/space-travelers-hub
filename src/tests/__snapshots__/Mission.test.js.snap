// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Mission Card 1`] = `
<tr
  id="jhgrfgw"
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  <td>
    <h3>
      To Mars
    </h3>
  </td>
  <td>
    Test missions card
  </td>
  <td>
    <div
      className="status-container"
      style={
        Object {
          "backgroundColor": null,
        }
      }
    >
      <h6>
        Active Member
      </h6>
    </div>
  </td>
  <td>
    <button
      className="mission-btn"
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid red",
          "color": "red",
        }
      }
      type="submit"
    >
      Leave Mission
    </button>
  </td>
</tr>
`;

exports[`Render Missions  1`] = `
<div>
  No missions
</div>
`;

exports[`Render Missions Page 1`] = `
<div>
  No missions
</div>
`;

exports[`render Missions  1`] = `
<div>
  No missions
</div>
`;
